added TopologyObject API

lilyminium opened the pull request:
[It'd be nice to have an easy way to add singular bonds/angles/dihedrals/impropers](https://github.com/MDAnalysis/mdanalysis/issues/2373#issuecomment-546520989). 

Changes made in this Pull Request:

**To universe.Universe:**
 - added add_TopologyObject/s
 - add_Bond/s is a convenience method
 - added delete_TopologyObject/s (have not added delete_Bond/s)
 - added refresh_fragments

**To topologyattrs._Connections:**
 - added delete_bonds
 - moved type checking to decorator

PR Checklist
------------
 - [x] Tests?
 - [ ] Docs?
    - What is the `types` keyword for in `_Connection.add_bonds`, and topology objects in general? Is `order` a reference to bond order -- what meaning does this have for angles?
 - [x] CHANGELOG updated?
 - [x] Issue raised/referenced?

By the way, I suppose you wouldn't be interested in 'tagging' TopologyObjects with extra information like force constants, etc. from topology files with that information?

richardjgowers commented on package/MDAnalysis/core/universe.py:
```diff
@@ -985,6 +986,208 @@ def add_Segment(self, **attrs):
         # return the new segment
         return self.segments[segidx]
+    
+    def add_TopologyObject(self, value, type=None, types=None, guessed=False,
```
I know they're called `TopologyObject`s, but I really don't like this method name.  I'd rather have just `add_Bonds` `add_Angles` and `add_Dihedrals`, I think they're clearer (and more likely to be found by someone).  Also I'd not bother with the singular version and just force people to give a 1 length list if they want to add a single Bond

lilyminium replied:
Changed -- this also simplifies the type checking and tests. I'm still not sure how to document the 'order' and 'types' keywords, or whether to even leave them in for non-Bonds (especially order).

lilyminium commented:
Hopefully fixes #2386. Added a _SymmetricConnection class for Bonds, Angles and Dihedrals with type checking and values where the first atom index is always smaller than the last. Added type checking to Impropers and a test to ensure that u.impropers returns impropers with the same indices as u._topology.impropers.

orbeckst commented:
> By the way, I suppose you wouldn't be interested in 'tagging' TopologyObjects with extra information like force constants, etc. from topology files with that information?

Do you mean similar to how we add Attributes in a flexible manner? Seems conceptually simple to add something like "force_constant", "equilibrium_value", "ryckaert_bellemans_coefficients", ... As long as these are "tags" and we don't really have to make sure that they are consistent between everything, it might be a convenient and extensible way to annotate with force field information. 

If we _want to be consistent_ then this gets harder...

lilyminium commented:
GROMACS also can have multiple definitions of the same bonds etc. that are additive, which complicates things. :(

lilyminium commented:
This last commit addresses #2392, but I don't know why it was written like that in the first place, so possibly not the best idea. There are more test offerings to please the codecov overlord.

orbeckst commented on package/MDAnalysis/core/topologyobjects.py:
```diff
@@ -539,18 +539,19 @@ def __init__(self, bondidx, universe, btype=None, type=None, guessed=None,
         nbonds = len(bondidx)
         # remove duplicate bonds
         if type is None:
-            type = np.repeat(None, nbonds).reshape(nbonds, 1)
```
If this is the fix for #2392 then I would separate it from this PR and make it a new one. It seems extra effort but it's much easier to discuss and review small things, especially if one is simple and the other complex.

lilyminium replied:
I created #2403 . Should the changes be removed from this PR? I think this PR was failing tests without it (how the bug was discovered in the first place)

orbeckst replied:
In principle, yes, remove. In practice, wait until #2403 is merged. Given that this PR will likely get squash-merged anyway, you can wait if any conflicts emerge.

orbeckst commented on package/MDAnalysis/core/universe.py:
```diff
@@ -992,6 +993,194 @@ def add_Segment(self, **attrs):
+        guessed : bool (optional)
+            Whether these connections have been guessed.
+        order : int (optional)
+            Bond order
```
The new methods should all come with a `.. versionadded:: 0.21.0`

orbeckst commented on package/MDAnalysis/core/universe.py:
```diff
@@ -992,6 +993,194 @@ def add_Segment(self, **attrs):
+            An iterable of: tuples of atom indices, or AtomGroups, 
+            or TopologyObjects. If every value is a TopologyObject, all 
+            keywords are ignored.
+        types : types (optional)
```
Is `types` a scalar, array-like, ... ? Not clear. (Also applies to docs below.)

It's especially confusing because `type` is also a Python thing.

lilyminium replied:
`types` can be either `None` or array-like. I think the LAMMPS parser actually uses them and I put them in ITPParser. When `types=None`, `TopologyObject.type` just returns the atom types of the atoms involved. `TopologyGroup.types()` returns an ordered unique list of every TopologyObject.type.

I'm not sure whether MDAnalysis uses the `type` information anywhere. 

```python
In [1]: import MDAnalysis as mda

In [2]: from MDAnalysis.tests.datafiles import *

In [3]: lmp = mda.Universe(LAMMPSdata)

In [4]: lmp.bonds[0].type
Out[4]: '3'

In [5]: lmp.bonds.types()
Out[5]: ['3', '5', '7', '2', '1', '4', '6']

In [6]: len(lmp.bonds)
Out[6]: 18336
```

orbeckst commented on package/CHANGELOG:
```diff
@@ -24,6 +24,8 @@ Fixes
   * Chainreader and continuous option work correctly when readers work for more 
     than one format (Issue #2353)
   * PDBQTParser now gives icode TopologyAttrs (Issue #2361)
+  * TopologyGroup no longer reshapes the type, guessed, and order properties
```
I'd make this a separate small PR (see below).

zemanj commented on package/MDAnalysis/core/universe.py:
```diff
@@ -992,6 +993,194 @@ def add_Segment(self, **attrs):
+        values : iterable of tuples, AtomGroups, or Bonds
+            An iterable of: tuples of atom indices, or AtomGroups, 
+            or Bonds. If every value is a Bond, all 
+            keywords are ignored.
```
It's not clear from the docstring how the method works with AtomGroups. You need to clarify that one needs to supply a list of AtomGroups with exactly two atoms per group. Maybe you could add an example for creating TIP4P water bonds such as
```python
import MDAnalysis as mda
from MDAnalysisTests.datafiles import GRO
u = mda.Universe(GRO)
ow_hw1 = u.select_atoms("resname SOL and (name OW or name HW1)").split('residue')
ow_hw2 = u.select_atoms("resname SOL and (name OW or name HW2)").split('residue')
ow_mw = u.select_atoms("resname SOL and (name OW or name MW)").split('residue')
u.add_bonds(ow_hw1 + ow_hw2 + ow_mw)
```
(or maybe there's a smarter way to do this)

zemanj commented on package/MDAnalysis/core/universe.py:
```diff
@@ -992,6 +993,194 @@ def add_Segment(self, **attrs):
+            or Bonds. If every value is a Bond, all 
+            keywords are ignored.
+        types : types (optional, default None)
+            Type of bond, or atom types
```
What are valid bond types?

zemanj commented on package/MDAnalysis/core/universe.py:
```diff
@@ -992,6 +993,194 @@ def add_Segment(self, **attrs):
+        attr.delete_bonds(values)
+
+    def delete_Bonds(self, values):
+        """Delete Bonds from this Universe and recreate fragments.
```
I think it's both irrelevant and confusing for the user that fragments are cleared.

zemanj commented on package/MDAnalysis/core/universe.py:
```diff
@@ -992,6 +993,194 @@ def add_Segment(self, **attrs):
+                           order=None):
+        """Add new TopologyObjects to this Universe
+
+        Parameters
```
Some parameters need more explanations:
What are valid values for `types` and `order`? Any restrictions?
What do they mean / what's their purpose?

What are the requirements for the different kinds of supported `values`?

richardjgowers replied:
@lilyminium I think order only makes sense for bonds (ie single/double), so keep it in this "private" method and only expose it in the `add_Bonds`

lilyminium commented on package/MDAnalysis/core/topologyattrs.py:
```diff
@@ -1739,9 +1727,52 @@ def add_bonds(self, values, types=None, guessed=True, order=None):
+        for i in sorted(idx, reverse=True):
+            del self.values[i]
+
+        for attr in ('types', '_guessed', 'order'): # faster
```
It was a note I left to remind myself that converting to an array was faster than deleting list values one-by-one -- removed now.

lilyminium commented on package/MDAnalysis/core/universe.py:
```diff
@@ -992,6 +993,194 @@ def add_Segment(self, **attrs):
         # return the new segment
         return self.segments[segidx]
+    
+    def _add_TopologyObjects(self, object_type, values, types=None, guessed=False,
```
Back when it was called `add_TopologyObject` the naming followed the `Universe.add_Residue`, `Universe.add_Segment`, `Universe.add_TopologyAttr`. Lower-casing it now because the method adds values to `Bonds` etc instead of adding a `Bonds` class (usually)

lilyminium commented on package/MDAnalysis/core/universe.py:
```diff
@@ -992,6 +993,194 @@ def add_Segment(self, **attrs):
+            except AttributeError:
+                types = None
+            guessed = first.is_guessed
+            order = first.order
```
I don't remember why I did that.... changed to take the attributes of each `TopologyObject`

lilyminium commented:
Thanks for the review, @zemanj!

richardjgowers commented on package/MDAnalysis/core/topologyattrs.py:
```diff
@@ -1908,17 +1959,43 @@ class Angles(_Connection):
+        return func(self, clean, *args, **kwargs)
+    return wrapper
 
 class Impropers(_Connection):
```
So impropers are also symmetrical see: https://github.com/MDAnalysis/mdanalysis/issues/2386#issuecomment-556033474

lilyminium replied:
Righto, switched them back -- we now can't assume the third atom is the central one that others are bonded to, as test_top.TOPBase used to.

richardjgowers commented on package/MDAnalysis/core/universe.py:
```diff
@@ -1030,6 +1270,19 @@ def _fragdict(self):
 
         return fragdict
+    
+    def clear_fragments(self):
```
Not sure this method is useful.  It shouldn't ever need to be called by a user (because it is to maintain internal book keeping) and it's a one liner anyway (`self._cache.pop('fragments', None)`)

richardjgowers commented on testsuite/MDAnalysisTests/data/adk_autopsf_formatted.pdb:
```diff
@@ -0,0 +1,3341 @@
+ATOM      1  N   MET 4   1     -11.053  26.680  12.742  0.00  0.00      4AKE  
```
Unless I missed something, these new files aren't used? (yet?)

lilyminium replied:
Huh, I have no idea how they got there.

lilyminium commented:
Thanks @richardjgowers!

lilyminium commented:
```python
self = <MDAnalysisTests.topology.test_top.TestPRMParser object at 0x7f7087f38860>
top = <MDAnalysis.core.topology.Topology object at 0x7f7087f4a390>

        def test_improper_atoms_bonded(self, top):
            vals = top.bonds.values
            for imp in top.impropers.values:
                for b in ((imp[0], imp[2]), (imp[1], imp[2]), (imp[2], imp[3])):
    >               assert (b in vals) or (b[::-1] in vals)
    E               assert ((31, 43) in [(10, 11), (10, 12), (4, 6), (4, 10), (0, 4), (25, 26), ...] or (43, 31) in [(10, 11), (10, 12), (4, 6), (4, 10), (0, 4), (25, 26), ...])

    testsuite/MDAnalysisTests/topology/test_top.py:127: AssertionError
```

Ah, this is the test that kicked off the symmetric improper dihedral ordering issue in #2386. It might be better to remove this test; if the central atom can be either first or third, you can no longer assume bonding anyway.

lilyminium commented:
Thanks for the review @zemanj -- added some new tests. 

> detection of duplicates with reversed order

These get removed before reaching the `_add_bonds` or `_delete_bonds` stage.

> BTW what happens if I try to add a TopologyGroup belonging to a different Universe?

The function only looks at the indices of the TopologyObjects involved, so it would add them unless they don't match the atom indices of the current Universe. Should this raise an error?

richardjgowers commented:
@lilyminium yeah I think a quick sanity check that the `.universe` attribute matches would probably save someone a headache one day.  There's not a `Universe.__eq__` method so you'll have to use `is`, it should raise an error, I think there's checks of this kind for adding AtomGroups.

zemanj commented:
@lilyminium @richardjgowers I'm not sure whether we should throw an error here. Imagine the case where you have two systems loaded in universes `u1`, `u2`, which are identical in terms of the molecules (atoms + bonds) they comprise but differ in terms of, say, angular potentials which are only present in `u1`.
In order to compare, e.g., the distributions of angles in these systems, it is very convenient to simply write `u2.add_angles(u1.angles)` and proceed from there.

richardjgowers commented:
@zemanj that only works for the case where the Universes are identically numbered (ie have an identical topology). I'm worried of a case that looks like:
```python
u1 = mda.Universe()
u2 = mda.Universe()
angles = u1.something()

# 1,000 lines of code later...

u2.add_angles(angles)
# now I'm confused when this doesn't work?
```

For the identical topology case I'd sooner make them use `u2.add_angles(u1.angles.to_indices())`

zemanj commented:
> For the identical topology case I'd sooner make them use `u2.add_angles(u1.angles.to_indices())`

Fair enough, that way one can still indirectly add TopologyObjects from different universes, and raising an error for direct use will hopefully be sufficient to prevent people from using the feature without thinking twice.

lilyminium commented:
> For the identical topology case I'd sooner make them use `u2.add_angles(u1.angles.to_indices())`

That doesn't seem too inconvenient. The code now ensures you cannot add or delete a TopologyGroup, list of TopologyObjects, or list of AtomGroups, from a different Universe.
